# locreadion
This is to remove the ambiguity in counting the number of reads mapped to regions, as some reads could be counted multiple times if they span cross regions or include splice sites.

## Requirements
- bedtools
- chromosome name starts with `chr`
- region files can be generated by [GTFanno](https://github.com/Oakento/GTFanno)

## Usage
```bash
locreadion -a <alignment> -r <region_dir> -o <output_dir>
```
